### ISSUE TITLE
Is there a way to draw concentric polygons/circles using large image? 

### ISSUE THREAD

**Issue by rwtam121** [girder/large_image#289]
Hi,

I was looking at the annotations documentation here: https://github.com/girder/large_image/blob/master/docs/annotations.md and was wondering if we could create a single polygon or a circle that has a polygon/circle within it. I ask because the nature of the annotations seems that if you put the points into the points array, a line will be drawn connecting the inner circle/polygon to the outer circle/polygon, like this:
![image](https://user-images.githubusercontent.com/29108723/41378792-1039c4c4-6f15-11e8-8ccc-d9584b02d2e7.png)

From documentation the points are specified like this:
    "points": [                      
      [5,6,0],                        
      [-17,6,0],
      [56,-45,6]
    ]
But is there any special way for it to support concentric polygons without a connecting line?

**Comment by manthey** [comment]
Internally, polygons can be have holes.  They can be specified like {"outer": [[5, 6, 0], [-17, 6, 0], [56, -45, 6]], "inner": [[[4, 5, 0], [2, 5, 0], [2, 1, 0]]]}, where the inner entry can have any number of lists, each of which is a separate hole.  There currently isn't any user interface for drawing a hole in a polygon, and you can't currently edit a polygon with holes.

From the user interface, how would you specify that you want to add a hole to a polygon?

**Comment by rwtam121** [comment]
Yeah, I saw that the large_image documentation didn't mention anything about creating holes. I was looking at GeoJSON's documentation and they noted that the polygons could be specified with holes, like what you said: With holes::

  { "type": "Polygon",
    "coordinates": [
      [ [100.0, 0.0], [101.0, 0.0], [101.0, 1.0], [100.0, 1.0], [100.0, 0.0] ],
      [ [100.2, 0.2], [100.8, 0.2], [100.8, 0.8], [100.2, 0.8], [100.2, 0.2] ]
      ]
   }

I was kind of hoping I would be able to add a list of more points underneath the "points" header below to put holes in my polygon, but as you said I don't think that's possible in large_image:
{
    "type": "polyline",                # Exact string.  Required
    <id, label, lineColor, lineWidth>  # Optional general shape properties
    "points": [                        # At least two points must be specified
      [5,6,0],                         # Coordinate.  At least two required
      [-17,6,0],
      [56,-45,6]
    ], **[[[more points here]]]**
    "closed": true,                    # Boolean.  Default is false.  Optional
    "fillColor": "rgba(0, 255, 0, 1)"  # String.  See note about colors.  Optional
}

Thanks for the help though.

**Comment by manthey** [comment]
I've add https://github.com/OpenGeoscience/geojs/issues/845 to eventually support this.

**Comment by rwtam121** [comment]
Had another question on this...is it possible to install the Geospatial plugin in Girder and be able to make GeoJSON formatted query statements in the web API (rather than using large_image formatted statements)? I ask because the Geospatial plugin says it "enables the storage and querying of GeoJSON formatted geospatial data" so does this mean I can switch from:

large_image format:
{
    "type": "polyline",                # Exact string.  Required
    <id, label, lineColor, lineWidth>  # Optional general shape properties
    "points": [                        # At least two points must be specified
      [5,6,0],                         # Coordinate.  At least two required
      [-17,6,0],
      [56,-45,6]
    ],
    "closed": true,                    # Boolean.  Default is false.  Optional
    "fillColor": "rgba(0, 255, 0, 1)"  # String.  See note about colors.  Optional
}

to 

GeoJSON format:
{
    "type": "Polygon",                
    "coordinates": [
      [ [100.0, 0.0], [101.0, 0.0], [101.0, 1.0], [100.0, 1.0], [100.0, 0.0] ],
      [ [100.2, 0.2], [100.8, 0.2], [100.8, 0.8], [100.2, 0.8], [100.2, 0.2] ]
      ],
}

I know you said there's no interface to support this, and this didn't work when I tried it, but just wondering if this was still an option.

**Comment by manthey** [comment]
The geospatial and large_image plugins use different data storage locations and types.

The large_image annotations are stored in a distinct Mongo collection (actually two collections: annotation and annotationelement).  The data in large_image annotations are their own schema which includes element types that don't exist in GeoJSON.

The Girder geospatial plugin handles GeoJSON data stored in a 'geo' field in the item collection, and searches only within that field.

**Comment by rwtam121** [comment]
I guess my question is that after I enabled the geospatial plugin and pressed rebuild and restart, 
![image](https://user-images.githubusercontent.com/29108723/41556249-03e0e184-72ef-11e8-8fcd-298fd77a7210.png)
![image](https://user-images.githubusercontent.com/29108723/41556257-0d213186-72ef-11e8-9b04-61782059a081.png)

I was kind of expecting the geospatial option in the REST API documentation, or an annotation schema that includes geoJSON data. But I don't see those options even after the plugin was installed. The screenshot below shows no geoJSON option:
![image](https://user-images.githubusercontent.com/29108723/41556397-80138356-72ef-11e8-8d30-f474cfb1d96b.png)

Am I expecting too much out of geoJSON, or is there something else I need to toggle to get it working? Sorry if the conversation went a little off topic from the large_image issue.

**Comment by dorukozturk** [comment]
@rwtam121 You would see the geospatial endpoints under "item" since that plugin is adding new routes to items. Just expand the item endpoints.

**Comment by rwtam121** [comment]
@dorukozturk Thanks for pointing me in the right direction! Sorry to keep moving off topic but I tried putting some geoJSON known to work in the geojson.io website (gets a rectangle as seen in the below left):
![image](https://user-images.githubusercontent.com/29108723/41565954-03abc7aa-730e-11e8-8d71-5aa297f6a07a.png)

and I went back to the REST API and put that json onto the image using PUT/item/{id}/geospatial.  
![image](https://user-images.githubusercontent.com/29108723/41566001-44098e5e-730e-11e8-889f-52983bc340b9.png)

But I got this error (TypeError: String indices must be integers):
![image](https://user-images.githubusercontent.com/29108723/41566045-83e056e8-730e-11e8-862c-f5c2a36e599b.png)

The field looks like it contains valid GeoJSON (it worked within geojson.io) so I am missing something fundamental here--should I use PUT/item/{id}/geospatial to put the overlay on the items? Thanks for the help!

**Comment by dorukozturk** [comment]
Sure, no problem. I think you only need the geometry part. Try it with the geometry portion of it. Something like:
```sh
{
        "type": "Polygon",
        "coordinates": [
          [
            [
              -86.484375,
              23.885837699862005
            ],
            [
              -73.47656249999999,
              23.885837699862005
            ],
            [
              -73.47656249999999,
              31.052933985705163
            ],
            [
              -86.484375,
              31.052933985705163
            ],
            [
              -86.484375,
              23.885837699862005
            ]
          ]
        ]
      }
```

**Comment by rwtam121** [comment]
Hmmm, I still got the same error when doing that:
TypeError: TypeError('string indices must be integers',)",
I'm wondering what could it mean by string indices? We have a list of coordinates and key-value pairs, but I don't think I'm indexing anything with the command...

**Comment by rwtam121** [comment]
Hi @dorukozturk , 

I was able to get the geospatial info from the coordinates with the help of https://github.com/girder/girder/issues/2744, using PUT /item/{id}/geospatial. But this just sets the geospatial info for the object:
![image](https://user-images.githubusercontent.com/29108723/41866122-13350b2c-7864-11e8-915d-c62dc42d6134.png)

I'm wondering we could translate that geospatial info into an overlay on top of that item, much like in the annotations case. Is that currently supported? I scanned the REST API and I don't think we can translate the geospatial info into an annotation.
